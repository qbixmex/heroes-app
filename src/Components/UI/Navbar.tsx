import React from 'react'
import { Link, NavLink, useNavigate } from 'react-router-dom'

export const Navbar = () => {
  const navigate = useNavigate();

  const handleLogout= () => {
    navigate("/login", {
      replace: true
    });
  };

  return (
    <nav className="navbar navbar-expand-sm navbar-dark bg-dark bg-gradient px-4 mb-4">
      <Link className="navbar-brand" to="/">
        Asociaciones
      </Link>

      <div className="navbar-collapse">
        <div className="navbar-nav">
          <NavLink
            className={ ({ isActive }) => `nav-item nav-link ${ isActive ? "active" : ""}`}
            to="/marvel"
          >
            Marvel
          </NavLink>

          <NavLink
            className={ ({ isActive }) => `nav-item nav-link ${ isActive ? "active" : ""}`}
            to="/dc"
          >
            DC
          </NavLink>
        </div>
      </div>

      <div className="navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end">
        <ul className="navbar-nav mr-auto">
          <li className='nav-item'>
            <button className="nav-link btn text-info">Daniel Gonz√°lez</button>
          </li>
          <li className="nav-item">
            <button
              className="nav-link btn"
              onClick={ handleLogout }
            >
              Logout
            </button>
          </li>
        </ul>
      </div>
    </nav>
  );
}