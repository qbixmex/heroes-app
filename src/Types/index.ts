export type AuthState = {
  name?: string;
  logged?: boolean;
};

export type AuthAction =
  | { type: types.login, payload: AuthState }
  | { type: types.logout };

export enum types {
  login = "[auth] Login",
  logout = "[auth] Logout"  
}
