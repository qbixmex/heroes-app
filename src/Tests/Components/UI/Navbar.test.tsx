import { mount } from 'enzyme';
import { AuthContext } from '../../../Auth/authContext';
import { MemoryRouter, Routes, Route } from 'react-router-dom';

import { Navbar } from '../../../Components/UI/Navbar';
import { types } from '../../../Types';

const mockNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate
}));

// @ts-ignore
describe("Tests on Navbar", () => {
  
  const contextValue = {
    user: {
      name: "John Doe",
      logged: true,
    },
    dispatch: jest.fn()
  };

  const wrapper = mount(
    <AuthContext.Provider value={ contextValue }>
      <MemoryRouter initialEntries={ ["/"] }>
        <Routes>
          <Route path="/" element={ <Navbar /> } />
        </Routes>
      </MemoryRouter>
    </AuthContext.Provider>
  );

  test("Should match with snapshot", () => {
    const username = wrapper.find(".text-info").text().trim()

    expect( wrapper ).toMatchSnapshot();
    expect( username ).toBe( contextValue.user.name );
  });

  test("Should call logout and call navigate with dispatch with arguments", () => {
    const mockEvent = {} as React.MouseEvent<HTMLButtonElement>;
    wrapper.find("#logout").prop<React.MouseEventHandler>("onClick")(mockEvent);

    expect( contextValue.dispatch ).toHaveBeenCalledWith({ type: types.logout });
    expect( mockNavigate ).toHaveBeenCalledWith( "/login", { "replace": true } );
  });

});
