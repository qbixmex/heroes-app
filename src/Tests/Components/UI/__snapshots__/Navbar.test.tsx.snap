// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on Navbar Should match with snapshot 1`] = `
<nav
  className="navbar navbar-expand-sm navbar-dark bg-dark bg-gradient px-4 mb-4"
>
  <a
    className="navbar-brand"
    href="/"
    onClick={[Function]}
  >
    Asociaciones
  </a>
  <div
    className="navbar-collapse"
  >
    <div
      className="navbar-nav"
    >
      <a
        className="nav-item nav-link "
        href="/marvel"
        onClick={[Function]}
      >
        Marvel
      </a>
      <a
        className="nav-item nav-link "
        href="/dc"
        onClick={[Function]}
      >
        DC
      </a>
      <a
        className="nav-item nav-link "
        href="/search"
        onClick={[Function]}
      >
        Search
      </a>
    </div>
  </div>
  <div
    className="navbar-collapse collapse w-100 order-3 dual-collapse2 d-flex justify-content-end"
  >
    <ul
      className="navbar-nav mr-auto"
    >
      <li
        className="nav-item"
      >
        <button
          className="nav-link btn text-info"
        >
          John Doe
        </button>
      </li>
      <li
        className="nav-item"
      >
        <button
          className="nav-link btn"
          id="logout"
          onClick={[Function]}
        >
          Logout
        </button>
      </li>
    </ul>
  </div>
</nav>
`;
